File: src/main/java/com/jeffreyorazulike/simpletron/memory/MemoryImpl.java
/*
 * Copyright (C) 2021 Jeffrey Orazulike
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package com.jeffreyorazulike.simpletron.memory;

/**
 *
 * @author Jeffrey Orazulike
 * @apiNote A concrete implementation of the {@link Memory} interface having
 * {@code 1000} blocks of memory
 */
public class MemoryImpl implements Memory {

    /**
     * Represents the memory
     */
    private final int memory[];

    private final int MEMORY_SIZE = 1000;
    private final int MAX_WORD = MEMORY_SIZE * 100 - 1;
    private final int MIN_WORD = MAX_WORD * -1;

    public MemoryImpl() {
        memory = new int[MEMORY_SIZE];
    }

    @Override
    public void set(int index, int value) throws ArrayIndexOutOfBoundsException {
        memory[index] = value;
    }

    @Override
    public int get(int index) throws ArrayIndexOutOfBoundsException {
        return memory[index];
    }

    @Override
    public int getConstant(Constant constant) {
        switch (constant) {
            case MEMORY_SIZE:
                return MEMORY_SIZE;
            case MAX_WORD:
                return MAX_WORD;
            case MIN_WORD:
                return MIN_WORD;
            default:
                throw new AssertionError(constant.name());
        }
    }
}

File: core/impl/src/main/kotlin/com/logickoder/simpletron/core/impl/component/MemoryImpl.kt
package com.logickoder.simpletron.core.impl.component

import com.jeffreyorazulike.simpletron.core.impl.utils.stopValue
import com.logickoder.simpletron.core.component.Memory

/**
 *
 * @author Jeffrey Orazulike [chukwudumebiorazulike@gmail.com]
 * Created on 03 at 5:03 PM
 *
 * A concrete implementation of the [Memory] having 1000 blocks of memory
 */
class MemoryImpl : com.logickoder.simpletron.core.component.Memory(1000) {

    /**
     * Represents the memory
     */
    private val memory = FloatArray(size)

    @Throws(ArrayIndexOutOfBoundsException::class, IllegalArgumentException::class)
    override fun set(address: Number, value: Number) = with(value.toFloat()) {
        if (this != stopValue())
            require(this in minWord..maxWord) {
                "$this is not in the range of $minWord to $maxWord"
            }
        memory[address.toInt()] = this
    }

    @Throws(ArrayIndexOutOfBoundsException::class)
    override fun get(address: Number): Float {
        return memory[address.toInt()]
    }
}